### PR TITLE
Make progress bar available with n_jobs!=1

### PR DESCRIPTION
**Pull request by masap**
## Motivation
Make progress bar available with ```n_jobs!=1```.

## Description of the changes
This PR fixes issue #2957.

Signed-off-by: Masashi Honma <masashi.honma@gmail.com>

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@toshihikoyanase @himkt Could you review this PR?

@xadrianzetx Please feel free to review this PR if you have time as an author of #3115.

**Comment by HideakiImamura** [comment]
@masap Let me share the useful information from @xadrianzetx. You can check https://github.com/optuna/optuna/pull/3115#discussion_r760397095 to try the progress bar with `n_jobs!=1`.

**Comment by masap** [comment]
@HideakiImamura Thanks! I will try it tomorrow.

**Comment by masap** [comment]
> Let me share the useful information from @xadrianzetx. You can check #3115 (comment) to try the progress bar with n_jobs!=1.

Great ! It works with my patch also.

**Comment by masap** [comment]
I fixed the PR. The previous patch made all processes outputting a progrress bar to a single ```stderr```. Since each process may show the progress ```100%``` twice, counting ```100%``` in single ```stderr``` could not find out a process which did now show ```100%```. The new patch will check if ```100%``` is displayed for each process.

**Comment by masap** [comment]
I am facing a RDB stroage failure. But I can not reproduce it in my local environment even if I use same python version... So I am testing on this CI. Sorry for frequent resubmit.

**Comment by masap** [comment]
These tests are running more than 2 hours unexpectedly. It runs about 3minutes in ordinary case. Could anybody stop the jobs ?

```
Tests (RDB storage) / tests-rdbstorage (3.6) (pull_request)
Tests (RDB storage) / tests-rdbstorage (3.7) (pull_request)
```

**Comment by himkt** [comment]
No worries @masap and I force stopped jobs.
As a side note, you can run CIs with your forked repo (open pull request from the forked repo to the forked repo).
Of course, there is no problem to run CIs on this PR but you can cancel jobs with your permission if jobs run on the forked repo.

**Comment by masap** [comment]
@himkt Thank you! I will try it on my repo.

**Comment by HideakiImamura** [comment]
@nzw0301 Could you review this PR instead of @himkt? I discussed it with him privately and came to the conclusion that it would be better to change reviewers.

**Comment by nzw0301** [comment]
sure.

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 1]
```diff
@@ -26,13 +29,33 @@ def objective(trial: optuna.Trial) -> float:
     return f(x, y)
 
 
-def run_optimize(args: Tuple[str, str]) -> None:
+def run_optimize(args: Tuple[str, str, int, bool]) -> str:
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:32]: I prefer defining another function for the progress bar to extend this `run_optimize`. What do you think?

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 2]
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
...
 
     study = optuna.load_study(study_name=study_name, storage=storage_url)
 
     trials = study.trials
     assert len(trials) == n_workers * 20
 
     _check_trials(trials)
+
+
+def test_multiprocess_with_progbar(
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:143]:
```suggestion
@pytest.mark.parametrize("n_jobs", [1, 4])
def test_multiprocess_with_progbar(
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 3]
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
...
+    capsys: _pytest.capture.CaptureFixture, storage_url: str
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:144]:
```suggestion
    capsys: _pytest.capture.CaptureFixture, storage_url: str, n_jobs: int
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 4]
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
...
+) -> None:
+    with capsys.disabled():
+        n_workers = 8
+        study_name = _STUDY_NAME
+        optuna.create_study(storage=storage_url, study_name=study_name)
+        with Pool(n_workers) as pool:
+            ret = pool.map(run_optimize, [(study_name, storage_url, 1, True)] * n_workers)
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:151]:
```suggestion
            ret = pool.map(run_optimize, [(study_name, storage_url, n_jobs, True)] * n_workers)
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 5]
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
...
+
+        study = optuna.load_study(study_name=study_name, storage=storage_url)
+
+        trials = study.trials
+        assert len(trials) == n_workers * 20
+
+        _check_trials(trials)
+
+
+def test_multiprocess_with_progbar_parallel(
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:166]: We can merge this function to `test_multiprocess_with_progbar` with `parametrize` decorator.

**Review by nzw0301 — CHANGES_REQUESTED** [review]

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 1 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Ok, I defined another function.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 2 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py:147]: Ok, fixed.

**Comment by masap** [comment]
In addition, I fixed a warning from black to avoid CI failure.

**Comment by masap** [comment]
```tox black``` warned for ```optuna/integration/xgboost.py``` in my local but CI does not allow the change. So I rolled back it.
```
    black run-test: commands[0] | black . --check --diff
    --- optuna/integration/xgboost.py       2021-12-13 23:34:01.614238 +0000
    +++ optuna/integration/xgboost.py       2021-12-14 00:17:52.997456 +0000
    @@ -72,10 +72,11 @@
                     message = "Trial was pruned at iteration {}.".format(epoch)
                     raise optuna.TrialPruned(message)
                 # The training should not stop.
                 return False
    
    +
     elif _imports.is_successful():
    
         def _get_callback_context(env: "xgb.core.CallbackEnv") -> str:
             """Return whether the current callback context is cv or train.
    
    @@ -111,10 +112,11 @@
                 self._trial.report(current_score, step=env.iteration)
                 if self._trial.should_prune():
                     message = "Trial was pruned at iteration {}.".format(env.iteration)
                     raise optuna.TrialPruned(message)
    
    +
     else:
    
         class XGBoostPruningCallback:  # type: ignore
             __doc__ = _doc
    
    would reformat optuna/integration/xgboost.py
    Oh no! 💥 💔 💥
    1 file would be reformatted, 281 files would be left unchanged.
```

**Comment by nzw0301** [comment]
## `black`

`black` in your environment is supposed to be a little bit old.

**Comment by masap** [comment]
The ```black``` in my environment is 21.12b0, the same as CI. I will check my environment. Anyway I fixed my PR, could you review this ?

**Comment by nzw0301** [comment]
I'm not familiar with logging of tqdm, which is used for the progress bar, but I can simplify the code as follows. 

```python
def run_optimize_with_progbar(args: Tuple[str, str, int]) -> str:
    study_name = args[0]
    storage_url = args[1]
    n_jobs = args[2]

    # Create a study
    study = optuna.load_study(study_name=study_name, storage=storage_url)

    tmp = tempfile.NamedTemporaryFile()

    with open(tmp.name, "w") as f:
        with contextlib.redirect_stderr(f):
            study.optimize(
                objective,
                n_trials=20,
                n_jobs=n_jobs,
                show_progress_bar=True,
            )
            f.flush()

    with open(tmp.name, "r") as f:
        return f.read()


@pytest.mark.parametrize("n_jobs", [1, 4])
def test_multiprocess_with_progbar(storage_url: str, n_jobs: int) -> None:
    n_workers = 8
    study_name = _STUDY_NAME
    optuna.create_study(storage=storage_url, study_name=study_name)
    with Pool(n_workers) as pool:
        progress_bars_as_str = pool.map(
            run_optimize_with_progbar, [(study_name, storage_url, n_jobs)] * n_workers
        )

        n_completed_bars = 0
        for progress_bar in progress_bars_as_str:
            if "20/20" in progress_bar and "100%" in progress_bar:
                n_completed_bars += 1

        assert n_completed_bars == n_workers

    study = optuna.load_study(study_name=study_name, storage=storage_url)

    trials = study.trials
    assert len(trials) == n_workers * 20

    _check_trials(trials)

```

What do you think?

**Comment by masap** [comment]
Thanks, Fixed. I have thought that we need ```capsys.disabled()``` to redirect the ```stderr```.

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 6]
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
...
     _check_trials(trials)
+
+
+@pytest.mark.parametrize("n_jobs", [1, 4])
+def test_multiprocess_with_progbar(storage_url: str, n_jobs: int) -> None:
+    n_workers = 8
+    study_name = _STUDY_NAME
+    optuna.create_study(storage=storage_url, study_name=study_name)
+    with Pool(n_workers) as pool:
+        results = pool.map(
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:145]: `results` is a little bit too general. How about this?
```suggestion
        progress_bars_as_str = pool.map(
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 7]
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
...
+            run_optimize_with_progbar, [(study_name, storage_url, n_jobs)] * n_workers
+        )
+        comp_cnt = 0
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:148]: This kinda abbreviation is also a little bit unclear for those who read this code later. How about this?
```suggestion
        n_completed_progbars = 0
```

**Review by nzw0301 — COMMENTED** [review]
Awesome. I've left minor comments. Could you take a look at these?

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 6 above)
Comment (nzw0301) [reply tests/storages_tests/rdb_tests/test_with_server.py:145]: Sorry, this variable was unclear (retuned value is not str). I'll revise the comment above.

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 8]
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
...
+        for result in results:
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:149]: Depending on the returned value by `pool.map`, but similarity `result` is not clear. One alternative might be as follows.

```suggestion
        for progress_bar in progress_bars:
```

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 6 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Ok, fixed.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 7 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Ok, fixed.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 8 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Thanks, all was fixed.

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 9]
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
     study.optimize(objective, n_trials=20)
 
 
+def run_optimize_with_progbar(args: Tuple[str, str, int]) -> str:
+    study_name = args[0]
+    storage_url = args[1]
+    n_jobs = args[2]
+    # Create a study
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:44]:
```suggestion

    # Create a study
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 10]
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
     study.optimize(objective, n_trials=20)
 
 
+def run_optimize_with_progbar(args: Tuple[str, str, int]) -> str:
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:40]: `run_optimize_with_progbar` returns progress bar as `str`, so how about this?

```suggestion
def get_completed_progbar(args: Tuple[str, str, int]) -> str:
```

If there exist a better name, please let me know!

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 11]
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
...
+    # Create a study
+    study = optuna.load_study(study_name=study_name, storage=storage_url)
+
+    tmp = tempfile.NamedTemporaryFile()
+    with open(tmp.name, "w") as f:
+        with contextlib.redirect_stderr(f):
+            # Run optimization
+            study.optimize(objective, n_trials=20, n_jobs=n_jobs, show_progress_bar=True)
+            f.flush()
+    with open(tmp.name, "r") as f:
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:53]:
```suggestion

    with open(tmp.name, "r") as f:
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 12]
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
...
+    study_name = args[0]
+    storage_url = args[1]
+    n_jobs = args[2]
+    # Create a study
+    study = optuna.load_study(study_name=study_name, storage=storage_url)
+
+    tmp = tempfile.NamedTemporaryFile()
+    with open(tmp.name, "w") as f:
+        with contextlib.redirect_stderr(f):
+            # Run optimization
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:50]: This sounds obvious:
```suggestion
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 13]
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
...
 
     _check_trials(trials)
+
+
+@pytest.mark.parametrize("n_jobs", [1, 4])
+def test_multiprocess_with_progbar(storage_url: str, n_jobs: int) -> None:
+    n_workers = 8
+    study_name = _STUDY_NAME
+    optuna.create_study(storage=storage_url, study_name=study_name)
+    with Pool(n_workers) as pool:
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:144]:
```suggestion

    with Pool(n_workers) as pool:
```

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 14]
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
...
+        progress_bars = pool.map(
+            run_optimize_with_progbar, [(study_name, storage_url, n_jobs)] * n_workers
+        )
+        n_completed_progbars = 0
```
Comment (nzw0301) [inline tests/storages_tests/rdb_tests/test_with_server.py:148]:
```suggestion

        n_completed_progbars = 0
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
let me leave additional minor comments for readability.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 9 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py:44]: I don't think that blank line is necessary because existing ```run_optimize``` does not have blank line before comment ```# Create a study```.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 10 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py:40]: The functionality of the ```run_optimize_with_progbar``` function is not only to return the progress bar, but also to run optimize. In fact, not only the progress bar but also the generated trials are checked subsequently. Therefore, the ```run_optimize_with_progbar``` function is the ```run_optimize``` function plus the ability to return the progress bar, and I think the name is appropriate.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 11 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py:54]: Ok, fixed.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 12 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py:50]: I added this comment because existing ```run_optimize``` has same comment before ```study.optimize```.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 13 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py:145]: I don't think that blank line is necessary because existing ```test_multiprocess``` does not have blank line before ```with Pool(n_workers) as pool:```.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 14 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py:150]: Ok, fixed.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 10 above)
Comment (nzw0301) [reply tests/storages_tests/rdb_tests/test_with_server.py:40]: I respectfully disagree with you. This function aims to get a completed progress bar that can be obtained by only using `optimize` method in Optuna. Since similar method `run_optimize` does return `None`, we expect `run_optimize_with_progbar` does also the same thing with progress bar option; we expect this returns `None`. Therefore the current name can mislead. 

cc: @toshihikoyanase

Comment (nzw0301) [reply tests/storages_tests/rdb_tests/test_with_server.py:40]: In terms of the test of optimisation results, I'm not sure this test is really necessary because the same test without a progress bar is done on another test.

**Comment by nzw0301** [comment]
We can merge this PR with [the squash option](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests), so please reflect your comments without the force option because we cannot see the difference from the previous commit by a new commit.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 12 above)
Comment (nzw0301) [reply tests/storages_tests/rdb_tests/test_with_server.py:50]: The existing code might not be a ground truth; personally, `run_optimize`'s comment looks also unnecessary.

**Comment by nzw0301** [comment]
Thank you for your quick update and response! If you are not sure whether or not to reflect my comments, you can wait for another reviewer's comments without hesitation!

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 10 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Ok, renamed to ```get_completed_progbar``` and removed check of trials.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 12 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Ok, removed.

**Comment by masap** [comment]
> We can merge this PR with the squash option, so please reflect your comments without the force option because we cannot see the difference from the previous commit by a new commit.

I have split this fix as a commit named "Additional fix for PR #3138". I needed to push with force option because I rebased my repo with current master.

**Comment by nzw0301** [comment]
I'll review this PR again after https://github.com/optuna/optuna/pull/3115 is merged into the master branch.

**Comment by nzw0301** [comment]
According to https://github.com/optuna/optuna/pull/3115#discussion_r759996119, `n_jobs != 1` with `timeout`-based progress bar does not work as we expected. So I'll suggest disabling the progress bar if progress bar tries to use `timeout` when `n_jobs != 1` in this PR.

**Comment by masap** [comment]
> disabling the progress bar if progress bar tries to use timeout when n_jobs != 1 in this PR.

Ok, fixed.

File: optuna/study/_optimize.py [hunk 15]
```diff
@@ -56,8 +56,11 @@ def _optimize(
     if not study._optimize_lock.acquire(False):
         raise RuntimeError("Nested invocation of `Study.optimize` method isn't allowed.")
 
-    # TODO(crcrpar): Make progress bar work when n_jobs != 1.
-    progress_bar = pbar_module._ProgressBar(show_progress_bar and n_jobs == 1, n_trials, timeout)
+    if show_progress_bar and timeout is not None and n_jobs != 1:
```
Comment (nzw0301) [inline optuna/study/_optimize.py:59]:
```suggestion
    if show_progress_bar and n_trials is None and timeout is not None and n_jobs != 1:
```

Comment (nzw0301) [inline optuna/study/_optimize.py:59]: When `n_trials` is not `None`, progress bar uses `n_trials` instead of `timeout`:
```suggestion
    if show_progress_bar and n_trials is None and timeout is not None and n_jobs != 1:
```

File: optuna/study/_optimize.py [hunk 16]
```diff
@@ -56,8 +56,11 @@ def _optimize(
...
+        warnings.warn("Progress bar is disabled because timeout is enabled and n_jobs != 1.")
```
Comment (nzw0301) [inline optuna/study/_optimize.py:60]: Like this?
```suggestion
        warnings.warn("The timeout-based progress bar is not supported with n_jobs != 1.")
```

File: tests/study_tests/test_study.py [hunk 17]
```diff
@@ -799,9 +799,18 @@ def _objective(trial: Trial) -> float:
 def test_optimize_with_progbar_parallel(capsys: _pytest.capture.CaptureFixture) -> None:
 
     study = create_study()
-    with pytest.warns(UserWarning, match="Progress bar only supports serial execution"):
-        study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
```
Comment (nzw0301) [inline tests/study_tests/test_study.py:803]: Could you add similar test to this part for `timeout`-based progress bar with `n_jobs!=1`?

**Review by nzw0301 — CHANGES_REQUESTED** [review]

File: optuna/study/_optimize.py (hunk 15 above)
Comment (masap) [reply optuna/study/_optimize.py]: Ok ,fixed.

File: optuna/study/_optimize.py (hunk 16 above)
Comment (masap) [reply optuna/study/_optimize.py]: Ok ,fixed.

File: tests/study_tests/test_study.py (hunk 17 above)
Comment (masap) [reply tests/study_tests/test_study.py:803]: Ok ,fixed.

**Review by nzw0301 — APPROVED** [review]
LGTM. Thanks!

File: tests/storages_tests/rdb_tests/test_with_server.py [hunk 18]
```diff
@@ -115,3 +134,20 @@ def test_multiprocess(storage_url: str) -> None:
...
+    with Pool(n_workers) as pool:
+        progress_bars = pool.map(
+            get_completed_progbar, [(study_name, storage_url, n_jobs)] * n_workers
+        )
+
+        n_completed_progbars = 0
+        for progress_bar in progress_bars:
+            if "20/20" in progress_bar and "100%" in progress_bar:
+                n_completed_progbars += 1
+        assert n_completed_progbars == n_workers
```
Comment (toshihikoyanase) [inline tests/storages_tests/rdb_tests/test_with_server.py:153]: Can we remove this test case?

1. `tests/storages_tests/rdb_tests/test_with_server.py` has the test cases to check the behavior of `RDBStorage` with multiple RDB servers such as MySQL and PostgreSQL. I don't think the progress bar feature depends on them.

2. The suggested test code executes `study.optimize` in parallel processes,  so each progress bar does not interact with others. So, I think we can simplify this test case by just checking the progress bar in a process.

File: tests/study_tests/test_study.py [hunk 19]
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
...
+    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
```
Comment (toshihikoyanase) [inline tests/study_tests/test_study.py:802]: This is not actually a change of this PR, but I think `n_jobs=-1` makes the test environmental dependent. How about setting `n_jobs=2` as can be seen in `test_optimize_with_progbar_parallel_timeout`?
```suggestion
    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=2)
```

File: tests/study_tests/test_study.py [hunk 20]
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
...
+    _, err = capsys.readouterr()
+
+    # search for progress bar elements in stderr
```
Comment (toshihikoyanase) [inline tests/study_tests/test_study.py:805]: A nit.
```suggestion
    # Search for progress bar elements in stderr.
```

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
I'm sorry for the delayed response. The change in `_optimize.py` looks great to me, but I have some comments on the test code.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 18 above)
Comment (masap) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Ok, removed.

File: tests/study_tests/test_study.py (hunk 19 above)
Comment (masap) [reply tests/study_tests/test_study.py]: Ok, fixed.

File: tests/study_tests/test_study.py (hunk 20 above)
Comment (masap) [reply tests/study_tests/test_study.py]: Ok, fixed both `test_optimize_with_progbar` and `test_optimize_with_progbar_parallel`.

I rebased and squashed these to one commit.

File: tests/study_tests/test_study.py [hunk 21]
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
...
-        study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
+    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=2)
+    _, err = capsys.readouterr()
+
+    # Search for progress bar elements in stderr.
+    assert "10/10" in err
+    assert "100%" in err
 
+
+def test_optimize_with_progbar_parallel_timeout(capsys: _pytest.capture.CaptureFixture) -> None:
```
Comment (toshihikoyanase) [inline tests/study_tests/test_study.py:810]: I found that the first assertion (i.e., `assert "10/10" not in err` ) in this function is True even if the value of `n_jobs` is 1. How about replacing the assertions with `assert "|" not in err`? It can be seen in `test_optimize_progbar_no_constraints` and it ensures that any progress-bar-related messages are not shown.

```diff
-    assert "10/10" not in err
-    assert "100%" not in err
+    # Testing for a character that forms progress bar borders.
+    assert "|" not in err
```

File: tests/study_tests/test_study.py [hunk 22]
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
...
 
     study = create_study()
-    with pytest.warns(UserWarning, match="Progress bar only supports serial execution"):
-        study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
+    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=2)
+    _, err = capsys.readouterr()
+
+    # Search for progress bar elements in stderr.
+    assert "10/10" in err
+    assert "100%" in err
```
Comment (toshihikoyanase) [inline tests/study_tests/test_study.py:807]: I found that most of the lines in this test function are duplicated with `test_optimize_with_progbar`. If we apply `parametrize` to `test_optimize_with_progbar` in terms of `n_jobs`, we can reduce the code and improve the maintainability.

```python
@pytest.mark.parametrize("n_jobs", [1, 2])
def test_optimize_with_progbar(n_jobs: int, capsys: _pytest.capture.CaptureFixture) -> None:

    study = create_study()
    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=n_jobs)
    _, err = capsys.readouterr()

    # Search for progress bar elements in stderr.
    assert "10/10" in err
    assert "100%" in err
```

In addition, I also found some missing test cases, for example, (`n_jobs=2`, `n_trials=10`, `timeout=2`) and (`n_jobs=2`, `n_jobs=None`, `timeout=None`). So, please apply `parametrize` to the following test functions as well.

- `test_optimize_without_progbar`
- `test_optimize_progbar_n_trials_prioritized`
- `test_optimize_progbar_no_constraints`

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your update! I found some missing test cases, and I guess we can further improve the test coverage.

File: tests/storages_tests/rdb_tests/test_with_server.py (hunk 18 above)
Comment (toshihikoyanase) [reply tests/storages_tests/rdb_tests/test_with_server.py]: Thanks!

File: tests/study_tests/test_study.py (hunk 20 above)
Comment (toshihikoyanase) [reply tests/study_tests/test_study.py]: Thank you!

File: tests/study_tests/test_study.py (hunk 21 above)
Comment (masap) [reply tests/study_tests/test_study.py:840]: Right, fixed.

File: tests/study_tests/test_study.py (hunk 22 above)
Comment (masap) [reply tests/study_tests/test_study.py]: > I found that most of the lines in this test function are duplicated with test_optimize_with_progbar. If we apply parametrize to test_optimize_with_progbar in terms of n_jobs, we can reduce the code and improve the maintainability.

Ok, fixed.

> In addition, I also found some missing test cases, for example, (n_jobs=2, n_trials=10, timeout=2) and (n_jobs=2, n_jobs=None, timeout=None). So, please apply parametrize to the following test functions as well.

Ok, fixed.

| show_progress_bar | n_trials | timeout | n_jobs | covered by |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| False | None | None | 1 | test_optimize_without_progbar_no_constraints |
| False | None | None | not 1 | test_optimize_without_progbar_no_constraints |
| False | None | not None | 1 | test_optimize_without_progbar_timeout |
| False | None | not None | not 1 | test_optimize_without_progbar_timeout |
| False | not None | None | 1 | test_optimize_without_progbar |
| False | not None | None | not 1 | test_optimize_without_progbar |
| False | not None | not None | 1 | test_optimize_without_progbar_n_trials_prioritized |
| False | not None | not None | not 1 | test_optimize_without_progbar_n_trials_prioritized |
| True | None | None | 1 | test_optimize_progbar_no_constraints |
| True | None | None | not 1 | test_optimize_progbar_no_constraints |
| True | None | not None | 1 | test_optimize_with_progbar_timeout |
| True | None | not None | not 1 | test_optimize_with_progbar_parallel_timeout |
| True | not None | None | 1 | test_optimize_with_progbar |
| True | not None | None | not 1 | test_optimize_with_progbar |
| True | not None | not None | 1 | test_optimize_progbar_n_trials_prioritized |
| True | not None | not None | not 1 | test_optimize_progbar_n_trials_prioritized |

File: tests/study_tests/test_study.py (hunk 21 above)
Comment (toshihikoyanase) [reply tests/study_tests/test_study.py:840]: Thanks!

**Review by toshihikoyanase — APPROVED** [review]
Thank you for summarizing the test case in https://github.com/optuna/optuna/pull/3138#discussion_r790035789. The change looks great to me!

**Comment by toshihikoyanase** [comment]
The table in https://github.com/optuna/optuna/pull/3138#discussion_r790035789 may suggest that we can reduce the number of test functions and simplify the test logic. I'm not sure, but we can work on it after discussing the test policy in https://github.com/optuna/optuna/issues/2974.
